### PR TITLE
OLS-885: Catalog generation from konflux snapshot

### PR DESCRIPTION
**Pull request by raptorsun**
## Description

add a new script `snapshot_to_catalog.sh` to generate catalog index file from Konflux snapshot.

To generate a new catalog index using snapshot `ols-bnxm2` to the catalog file `lightspeed-catalog-4.16/index.yaml` for bundle `0.1.0`, execute this command.
```
./hack/snapshot_to_catalog.sh ols-bnxm2 lightspeed-catalog-4.16/index.yaml
```
## Type of change

- [ ] Refactor
- [ ] New feature
- [ ] Bug fix
- [ ] CVE fix
- [ ] Optimization
- [x] Documentation Update
- [x] Configuration Update
- [ ] Bump-up dependent library

## Related Tickets & Documents

- Related Issue # [OLS-855](https://issues.redhat.com//browse/OLS-855)
- Closes #

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] PR has passed all pre-merge test jobs.
- [ ] If it is a core feature, I have added thorough tests.

## Testing
- Please provide detailed steps to perform tests related to this code change.
- How were the fix/results from this change verified? Please provide relevant screenshots or results.

### CODE REVIEW

File: hack/snapshot_to_catalog.sh [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+cat <<EOF >>"${CATALOG_FILE}"
+---
+schema: olm.channel
+package: lightspeed-operator
+name: preview
+entries:
+  - name: lightspeed-operator.v${BUNDLE_TAG}
+EOF
+
+${OPM} validate $(dirname "${CATALOG_FILE}")
```
Comment (tisnik) [inline hack/snapshot_to_catalog.sh:132]: I'd quote $(dirname) to avoid splitting

File: hack/snapshot_to_catalog.sh [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  usage
+fi
+
+SNAPSHOT_REF=${1:-"ols-bnxm2"}
+CATALOG_FILE=${2:-"lightspeed-catalog-4.16/index.yaml"}
+BUNDLE_TAG=${3:-"0.1.0"}
+echo "update bundle ${BUNDLE_TAG} catalog ${CATALOG_FILE} from snapshot ${SNAPSHOT_REF}"
+
+# cache the snapshot from Konflux
+TMP_SNAPSHOT_JSON=$(mktemp)
```
Comment (tisnik) [inline hack/snapshot_to_catalog.sh:21]: in following code it is better to double quote it (in all places)

File: hack/snapshot_to_catalog.sh [hunk 3]
```diff
@@ -0,0 +1,138 @@
+#!/usr/bin/env bash
+
+usage() {
+  echo "Usage: $0 <snapshot-ref> <catalog-file> <bundle-tag>"
+  echo "  snapshot-ref: the snapshot reference to use, example: ols-bnxm2"
+  echo "  catalog-file: the catalog index file to update, example: lightspeed-catalog-4.16/index.yaml"
+  echo "  bundle-tag: the tag of the bundle to update, example: 0.1.0"
+  exit 1
```
Comment (xrajesh) [inline hack/snapshot_to_catalog.sh:8]: We will have the bundle-tag also in the snapshot right , we dont want the users to provide it 
Also we need to create seperate folders for all our konflux release catalog Dockerfile and the configs

File: hack/snapshot_to_catalog.sh (hunk 1 above)
Comment (raptorsun) [reply hack/snapshot_to_catalog.sh:132]: What kind of splitting dirname can give us?

Comment (raptorsun) [reply hack/snapshot_to_catalog.sh:132]: Do you mean dirname can return "(" and ")"?

File: hack/snapshot_to_catalog.sh (hunk 3 above)
Comment (raptorsun) [reply hack/snapshot_to_catalog.sh:7]: updated. we fetch bundle version from the OPM rendered bundle information.

File: hack/snapshot_to_catalog.sh (hunk 1 above)
Comment (raptorsun) [reply hack/snapshot_to_catalog.sh]: added a quote in the new commit, is that what you mean?

File: hack/snapshot_to_catalog.sh [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  echo "  snapshot-ref: the snapshot reference to use, example: ols-bnxm2"
+  echo "  catalog-file: the catalog index file to update, example: lightspeed-catalog-4.16/index.yaml"
+  exit 1
+}
+
+if [ $# == 0 ]; then
+  usage
+fi
+
+SNAPSHOT_REF=${1:-"ols-bnxm2"}
```
Comment (xrajesh) [inline hack/snapshot_to_catalog.sh:14]: @raptorsun Do we want a default option ? We want to force the user to provide snapshot ref .

Comment (raptorsun) [reply hack/snapshot_to_catalog.sh]: now updated: snapshot is required, only catalog file is default to the 4.16 catalog file

File: hack/snapshot_to_catalog.sh (hunk 2 above)
Comment (raptorsun) [reply hack/snapshot_to_catalog.sh:53]: all quoted ;)

File: hack/snapshot_to_catalog.sh [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+)
+
+echo "Catalog will use the following images:"
+echo "BUNDLE_IMAGE=${BUNDLE_IMAGE}"
+echo "OPERATOR_IMAGE=${OPERATOR_IMAGE}"
+echo "CONSOLE_IMAGE=${CONSOLE_IMAGE}"
+echo "SERVICE_IMAGE=${SERVICE_IMAGE}"
+
+echo BUNDLE_IMAGE_ORIGIN=${BUNDLE_IMAGE_ORIGIN}
+${OPM} render ${BUNDLE_IMAGE_ORIGIN} --output=yaml >"${TEMP_BUNDLE_FILE}"
```
Comment (xrajesh) [inline hack/snapshot_to_catalog.sh:142]: @raptorsun TEMP_BUNDLE_FILE is this the right name , the render output is catalog config right ? I am baffled by the naming . if not , this PR LGTM

Comment (raptorsun) [reply hack/snapshot_to_catalog.sh:142]: That TEMP_BUNDLE_FILE is  the bundle part of the catalog  that the `opm` renders. It is later assembled into the catalog file. Is a name like `TEMP_BUNDLE_PART_OF_CATALOG` better?

Comment (xrajesh) [reply hack/snapshot_to_catalog.sh:142]: Lets leave it as is  TEMP_BUNDLE_FILE confused me .

**Comment by raptorsun** [comment]
> @raptorsun We have to generate catalog/index per release now , can we update all the catalogs by default ?

yes, we can run the script for each catalog with the selected snapshot reference:
```shell
./hack/snapshot_to_catalog.sh ols-xxxxx lightspeed-catalog-4.16/index.yaml
./hack/snapshot_to_catalog.sh ols-yyyyy lightspeed-catalog-4.15/index.yaml
```

**Pull request merged**